main.c

```diff
@@ -1,142 +1,39 @@
-#include "eeprom/eeprom.h"
+#include "operation/operation.h"
 
-static const uint16_t stored_entry_index_address = I2C_MEMORY_SIZE-1;
-static const uint16_t stored_rotation_address = I2C_MEMORY_SIZE-2;
-
-int main() {
+int main()
+{
     stdio_init_all();
 
-    LED led_3 = {.pin = LED_3};
-    init_pwm(&led_3);
-
-    init_switch(SW_0);
-
+    oper_st state = init_operation();
+    LED led_3 = init_pwm(LED_3);
+    SW sw_proceed = init_switch(SW_0);
+    SW sw_log = init_switch(SW_1); // this btn for read log
     init_opto_fork();
     init_stepper();
     init_piezo();
 
-    initialize_eeprom();
-
-    int rotations = 0; // read from eeprom, zero by default
-    uint8_t current_entry_index = 0;
-
-    read_from_eeprom(stored_rotation_address, (uint8_t *)&rotations, 1);
-    if (rotations < 0 || rotations > 7) rotations = 0;
+    // print_state(state);
 
-    //write first Boot log to memory
-    read_from_eeprom(stored_entry_index_address, &current_entry_index, 1);
-    if(current_entry_index >= MAX_ENTRIES || current_entry_index < 0) current_entry_index = 0;
-
-    char boot_log[] = "Boot";
-//    printf("%s\n", boot_log);
-    write_log_entry(boot_log, &current_entry_index, stored_entry_index_address);
-
-    while (true) {
-        /// AT STARTUP ///
-        /// Minimum:
-        // Blink LED,
-        // Until button pressed
-        /// Advanced:
-        // "Remembers state across boot/power down":
-        // Log to EEPROM
-        // LoRaWAN Report: boot.
-
-        while (!switch_pressed_debounced(SW_0)) {
-            toggle_pwm(&led_3);
-            sleep_ms(500);
+    while (true)
+    {
+        if (state.current_comp_idx == PILLCOMP_COUNT)
+        {
+            blink_until_sw_pressed(&sw_proceed, &led_3, &state);
         }
 
-        set_led_brightness(&led_3, PWM_OFF);
+        calibrate(&state);
 
-        /// BUTTON PRESSED #1 ///
-        /// Minimum:
-        // Calibrate,
-        // 'At least one full "turn" (revolution?)',
-        // Stops at opto-fork
-        /// Advanced:
-        // Log to EEPROM
-        // Calibrate according to state:
-        // opposite direction so the pills aren't dispensed.
-        // LoRaWAN Report: Calibration started.
+        wait_until_sw_pressed(&sw_proceed, &led_3, &state);
 
-        calibrate(rotations);
+        dispense(&state, &led_3, &sw_log);
 
-        /// AFTER INITIAL CALIBRATION ///
-        /// Minimum:
-        // LED turns on
-        // [fill pill slots]
-        // Waits for button press
-        /// Advanced:
-        // Log to EEPROM
-        // ? Dispense pill if previous dispensing was not finished due to reboot ?
-        // LoRaWAN Report: Calibration finished.
+        // task 1: print read log  if btn pressed
 
-        set_led_brightness(&led_3, PWM_SOFT);
-
-        while (!switch_pressed_debounced(SW_0)) {
-            sleep_ms(50);
+        if (switch_pressed_debounced(&sw_log))
+        {
+            read_log_entry(MAX_ENTRIES);
         }
```

Comment:
And if we had the opportunity to read_log within the 3 waiting loops,
this would be redundant.

Also, this would require the user to be holding the button at the time the dispensing was finishing,
which I find unintuitive.